Use official k8s watcher to watch for k8s events

aanm opened the pull request:
Long story short: don't reinvent the wheel. Read per commit. ~Last commit can be seen as a bug fix, I can create a backport commit for that bug fix if anyone wants it.~

Fixes: #6252

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7301)
<!-- Reviewable:end -->

aanm committed:
pkg/k8s: use official k8s library to watch events from kube-apiserver

There no gain on using our own library to watch events from kubernetes
so we can simply remove our own library and use the official library
from client-go.

Signed-off-by: André Martins <andre@cilium.dev>

aanm committed:
k8s: remove resync from k8s watchers

Since re-sync does not resyncs all objects from kube-apiserver every N
minutes as originally thought we can simply set it to zero.

Signed-off-by: André Martins <andre@cilium.dev>

aanm committed:
k8s: handle objects that are deleted without an event received

If client-go loses connectivity with the kube-apiserver it can lose
delete events. Once client-go reconnects to kube-apiserver, and notices
those objects no longer exist in kube-apiserver, cache.Informer will
create delete events where the deleted structures will be the type of
cache.DeletedFinalStateUnknown. This struct contains the last known
state of the non-existing object from the client's POV.

Signed-off-by: André Martins <andre@cilium.dev>

joestringer reviewed (commented):
Janitors are requested for review here, but it looks like the agent/k8s teams could provide much better input. Acknowledging the PR but not reviewing at this stage.

aanm committed:
k8s: ignore k8s update events if the object is considered equal

As we have our own implementation for each object that we are listening
from k8s we ignore all updates where the objects are considered the
same.

Signed-off-by: André Martins <andre@cilium.dev>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -337,157 +337,194 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
 		return fmt.Errorf("Unable to create cilium network policy client: %s", err)
 	}
 
+	serKNPs := serializer.NewFunctionQueue(1024)
```
Let's make the 1024 configurable via an option.

tgraf commented on pkg/k8s/factory_functions_test.go:
```diff
@@ -25,85 +25,10 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-func (s *K8sSuite) Test_missingK8sNetworkPolicyV1(c *C) {
```
I assume this should be folded into the previous commit?

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -367,6 +367,10 @@ func (d *Daemon) EnableK8sWatcher() error {
 					if oldK8sNP := k8s.CopyObjToV1NetworkPolicy(oldObj); oldK8sNP != nil {
 						if newK8sNP := k8s.CopyObjToV1NetworkPolicy(newObj); newK8sNP != nil {
+							if k8s.EqualV1NetworkPolicy(oldK8sNP, newK8sNP) {
+								return
```
I think we need to cover this with a metric somehow to understand how much we are ignoring. Ideally we bump a metric for each event received here and have a label type = { invalid | equal | valid }

I have some ideas here. Happy to work on this after we merge this PR.

aanm committed:
k8s: add k8s watcher queue size as a CLI parameter

Signed-off-by: André Martins <andre@cilium.dev>
